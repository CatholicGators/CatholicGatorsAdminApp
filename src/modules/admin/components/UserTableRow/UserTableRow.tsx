import React from "react"
import {
  TableRow,
  TableCell,
  Checkbox,
  Avatar,
  Switch,
  Theme,
  createStyles,
  withStyles
} from "@material-ui/core";

const styles = (theme: Theme) => createStyles({
  hiddenColxs: {
    [theme.breakpoints.down('xs')]: {
      display: 'none'
    }
  },
  hiddenColsm: {
    [theme.breakpoints.down('sm')]: {
      display: 'none'
    }
  },
  profilePicCol: {
    paddingRight: theme.spacing.unit * 2
  }
})

type Props = {
  user: any,
  classes: any,
  isSelected: boolean,
  handleSelect: (id) => any,
  handleApproveToggle: (user: any, checked: boolean) => any,
  handleAuthorizeToggle: (user: any, checked: boolean) => any
}

export const UserTableRow = (props: Props) => {
  const {
    user,
    classes,
    isSelected,
    handleSelect,
    handleApproveToggle,
    handleAuthorizeToggle
  } = props

  return (
    <TableRow
      hover
      aria-checked={isSelected}
      tabIndex={-1}
      key={user.id}
      selected={isSelected}
    >
      <TableCell padding="checkbox">
        <Checkbox
          onClick={() => handleSelect(user.id)}
          checked={isSelected}
        />
      </TableCell>
      <TableCell className={classes ? classes.profilePicCol : null}>
        <Avatar src={user.data.photoURL} />
      </TableCell>
      <TableCell className={classes ? classes.hiddenColsm : null}>{user.data.name}</TableCell>
      <TableCell>{user.data.email}</TableCell>
      <TableCell className={classes ? classes.hiddenColxs : null}>
        <Switch
          checked={user.data.isApproved}
          onChange={(_, checked) => handleApproveToggle(user, checked)}/>
      </TableCell>
      <TableCell className={classes ? classes.hiddenColxs : null}>
        <Switch
          disabled={!user.data.isApproved}
          checked={user.data.isAdmin}
          onChange={(_, checked) => handleAuthorizeToggle(user, checked)}
        />
      </TableCell>
    </TableRow>
  );
}

export default withStyles(styles)(UserTableRow)
