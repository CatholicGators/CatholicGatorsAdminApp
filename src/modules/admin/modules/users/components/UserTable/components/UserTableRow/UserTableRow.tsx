import React from "react"
import { connect } from "react-redux"
import {
    TableRow,
    TableCell,
    Checkbox,
    Avatar,
    Switch,
    Theme,
    createStyles,
    withStyles,
    IconButton,
} from "@material-ui/core"
import CheckIcon from "@material-ui/icons/Check"
import { updateUser } from "../../../../redux/actions/usersActions"

export const styles = (theme: Theme) =>
    createStyles({
        hiddenxs: {
            [theme.breakpoints.down("xs")]: {
                display: "none",
            },
        },
        hiddensm: {
            [theme.breakpoints.down("sm")]: {
                display: "none",
            },
        },
        avatarIconBtn: {
            padding: 0,
            [theme.breakpoints.up("sm")]: {
                display: "none",
            },
        },
        profilePicCol: {
            paddingRight: theme.spacing(2),
        },
    })

type Props = {
    user: any
    classes: any
    isSelected: boolean
    handleSelect: (id) => any
    updateUser: (user) => void
}

export class UserTableRow extends React.Component<Props> {
    handleApproveToggle(user, checked) {
        this.props.updateUser({
            ...user,
            isApproved: checked,
        })
    }

    handleAuthorizeToggle(user, checked) {
        this.props.updateUser({
            ...user,
            isAdmin: checked,
        })
    }

    render() {
        const { user, classes, isSelected, handleSelect } = this.props

        return (
            <TableRow
                hover
                aria-checked={isSelected}
                tabIndex={-1}
                key={user.id}
                selected={isSelected}
            >
                <TableCell
                    className={classes ? classes.hiddenxs : null}
                    padding="checkbox"
                >
                    <Checkbox
                        onClick={() => handleSelect(user.id)}
                        checked={isSelected}
                    />
                </TableCell>
                <TableCell className={classes ? classes.profilePicCol : null}>
                    <Avatar
                        className={classes ? classes.hiddenxs : null}
                        src={user.photoUrl}
                    />
                    <IconButton
                        className={classes ? classes.avatarIconBtn : null}
                        onClick={() => handleSelect(user.id)}
                    >
                        <Avatar src={isSelected ? null : user.photoUrl}>
                            {isSelected ? <CheckIcon /> : null}
                        </Avatar>
                    </IconButton>
                </TableCell>
                <TableCell className={classes ? classes.hiddensm : null}>
                    {user.name}
                </TableCell>
                <TableCell>{user.email}</TableCell>
                <TableCell>
                    <Switch
                        id="approve-switch"
                        checked={user.isApproved}
                        onChange={(_, checked) =>
                            this.handleApproveToggle(user, checked)
                        }
                    />
                </TableCell>
                <TableCell>
                    <Switch
                        id="authorize-switch"
                        disabled={!user.isApproved}
                        checked={user.isAdmin}
                        onChange={(_, checked) =>
                            this.handleAuthorizeToggle(user, checked)
                        }
                    />
                </TableCell>
            </TableRow>
        )
    }
}

const mapDispatchToProps = (dispatch) => ({
    updateUser: (user) => dispatch(updateUser(user)),
})

export default connect(
    null,
    mapDispatchToProps
)(withStyles(styles)(UserTableRow))
