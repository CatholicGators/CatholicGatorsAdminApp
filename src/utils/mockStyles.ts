import { createMuiTheme } from '@material-ui/core'

export default (styles) => Object.keys(styles(createMuiTheme())).reduce((acc, curr) => ({...acc, [curr]: "test"}), {})