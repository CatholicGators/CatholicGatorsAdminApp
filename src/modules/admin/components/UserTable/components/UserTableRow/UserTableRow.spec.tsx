import React from 'react'
import { shallow } from 'enzyme';
import { UserTableRow, styles} from './UserTableRow';
import { Checkbox, IconButton } from '@material-ui/core';
import mockStyles from '../../../../../../utils/mockStyles';

describe('UserTableRow', () => {
    let wrapper, props;

    beforeEach(() => {
        props = {
            classes: mockStyles(styles),
            user: {
                id: 'test',
                photoUrl: ''
            },
            isSelected: true,
            handleSelect: jest.fn(),
            updateUser: jest.fn()
        }
        wrapper = shallow(<UserTableRow {...props}/>)
    })

    it('matches snapshot', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('calls handleSelect with user.id when the checkbox is clicked', () => {
        wrapper.find(Checkbox).simulate('click')

        expect(props.handleSelect).toHaveBeenCalledWith(props.user.id)
    })

    it('calls handleSelect with user.id when the Avatar is clicked', () => {
        wrapper.find(IconButton).simulate('click')

        expect(props.handleSelect).toHaveBeenCalledWith(props.user.id)
    })

    it('updates the user when the approve switched is toggled', () => {
        const checked = true
        wrapper.find('#approve-switch').simulate('change', {target: {checked}}, checked)

        expect(props.updateUser).toHaveBeenCalledWith({
            ...props.user,
            isApproved: checked
        })
    })

    it('updates the user when the authorize switch is toggled', () => {
        const checked = true
        wrapper.find('#authorize-switch').simulate('change', {target: {checked}}, checked)

        expect(props.updateUser).toHaveBeenCalledWith({
            ...props.user,
            isAdmin: checked
        })
    })
})