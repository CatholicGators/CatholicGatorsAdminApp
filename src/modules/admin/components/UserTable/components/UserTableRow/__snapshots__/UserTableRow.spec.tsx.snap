// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTableRow matches snapshot 1`] = `
<WithStyles(ForwardRef(TableRow))
  aria-checked={true}
  hover={true}
  key="test"
  selected={true}
  tabIndex={-1}
>
  <WithStyles(ForwardRef(TableCell))
    className="test"
    padding="checkbox"
  >
    <WithStyles(ForwardRef(Checkbox))
      checked={true}
      onClick={[Function]}
    />
  </WithStyles(ForwardRef(TableCell))>
  <WithStyles(ForwardRef(TableCell))
    className="test"
  >
    <WithStyles(ForwardRef(Avatar))
      className="test"
    />
    <WithStyles(ForwardRef(IconButton))
      className="test"
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(Avatar))
        src={null}
      >
        <CheckIcon />
      </WithStyles(ForwardRef(Avatar))>
    </WithStyles(ForwardRef(IconButton))>
  </WithStyles(ForwardRef(TableCell))>
  <WithStyles(ForwardRef(TableCell))
    className="test"
  />
  <WithStyles(ForwardRef(TableCell)) />
  <WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(Switch))
      id="approve-switch"
      onChange={[Function]}
    />
  </WithStyles(ForwardRef(TableCell))>
  <WithStyles(ForwardRef(TableCell))>
    <WithStyles(ForwardRef(Switch))
      disabled={true}
      id="authorize-switch"
      onChange={[Function]}
    />
  </WithStyles(ForwardRef(TableCell))>
</WithStyles(ForwardRef(TableRow))>
`;
